### DISCUSSION TITLE
Form fields not submitting when in accordion closed state, Does anyone know how to make it work ?

### DISCUSSION THREAD

**Discussion by luvv8766** [radix-ui/primitives#3911] [Help]
## Question

I put forms in radix accordion component but when Im trying to submit the form fields its not submitting my form data.
Does anyone knows how to make it work ?

**Comment by FredrikMWold** [comment]
If you open up the devtools and look at the accordion content when opening and closing an item you can see that the content gets unmounted when the item is closed. This is the reason why its not showing up in the form data becuase the form field does not exist at the time of submission.

Looking through the sourcecode it looks like the collaplsable primative only renders the content when it is open:

```js
return (
    <Primitive.div
      data-state={getState(context.open)}
      data-disabled={context.disabled ? '' : undefined}
      id={context.contentId}
      hidden={!isOpen}
      {...contentProps}
      ref={composedRefs}
      style={{
        [`--radix-collapsible-content-height` as any]: height ? `${height}px` : undefined,
        [`--radix-collapsible-content-width` as any]: width ? `${width}px` : undefined,
        ...props.style,
      }}
    >
      {isOpen && children} // only show children if isOpen is true
    </Primitive.div>
  );
```

I could not see any way, atleast not any trivial way to override it. So to fix your issue i think you need to manage your own form state and use that when submitting.

**Comment by g-a-v-i-n** [comment]
You can use forceMount on <Content />, but you'll have to manage open/closed state in a controlled fashion.

**Comment by DrummyFloyd** [comment]
even if i use `shadcn` the accordion components is based on this lib 

i resolved this with this class

```tsx
const AccordionContent = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Content>
>(({ className, children, ...props }, ref) => (
  <AccordionPrimitive.Content
    ref={ref}
     forceMount // This ensures the content is always mounted
    className={cn(
      "overflow-hidden text-sm transition-all",
      {
        "data-[state=closed]:hidden": "hidden", // Add this line to hide content when closed
        "data-[state=open]:block": "block",     // Add this line to show content when open
      },
      className
    )}
    {...props}
  >
    <div className={cn("pb-4 pt-0", className)}>{children}</div>
  </AccordionPrimitive.Content>
))
```

**Comment by mech** [comment]
I am using this in order to have animation work. I need forceMount for my React-Hook-Form to do validation.

```javascript
import { motion, useAnimation } from "framer-motion";
import { useEffect, useRef } from "react";

const AccordionMountedContent = ({ children, ...props }) => {
  const controls = useAnimation();
  const contentRef = useRef(null);
  const divRef = useRef(null);

  useEffect(() => {
    const content = contentRef.current;

    if (!divRef.current || !content) return;

    const updateStateAnimation = () => {
      const state = contentRef.current.getAttribute("data-state");

      if (state === "open") {
        controls.start({
          height: "var(--radix-accordion-content-height)",
          opacity: 1,
          transition: { duration: 0.3, ease: "easeInOut" },
        });
      } else if (state === "closed") {
        controls.start({
          height: 0,
          opacity: 0,
          transition: { duration: 0.3, ease: "easeInOut" },
        });
      }
    };

    updateStateAnimation();

    const resizeObserver = new ResizeObserver(() => {
      const currentHeight = divRef.current.clientHeight;
      content.style.cssText = `--radix-accordion-content-height: ${currentHeight}px;`;
    });

    const observer = new MutationObserver(updateStateAnimation);
    if (contentRef.current) {
      observer.observe(contentRef.current, { attributes: true });
    }

    resizeObserver.observe(divRef.current);

    return () => {
      observer.disconnect();
      resizeObserver.disconnect();
    };
  }, [controls]);

  return (
    <AccordionContent ref={contentRef} forceMount {...props}>
      <motion.div initial={{ height: 0 }} animate={controls}>
        <div ref={divRef}>{children}</div>
      </motion.div>
    </AccordionContent>
  );
};
```

**Comment by sabinayakc** [comment]
> even if i use `shadcn` the accordion components is based on this lib
> 
> i resolved this with this class
> 
> const AccordionContent = React.forwardRef<
>   React.ElementRef<typeof AccordionPrimitive.Content>,
>   React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Content>
> >(({ className, children, ...props }, ref) => (
>   <AccordionPrimitive.Content
>     ref={ref}
>      forceMount // This ensures the content is always mounted
>     className={cn(
>       "overflow-hidden text-sm transition-all",
>       {
>         "data-[state=closed]:hidden": "hidden", // Add this line to hide content when closed
>         "data-[state=open]:block": "block",     // Add this line to show content when open
>       },
>       className
>     )}
>     {...props}
>   >
>     <div className={cn("pb-4 pt-0", className)}>{children}</div>
>   </AccordionPrimitive.Content>
> ))

the only problem with this approach is animation. How do you deal with animation? Currently its very glitchy.

**Comment by MendyLanda** [comment]
```ts
export function FormWithAccordionErrors() {
  const [accordionOpen, setAccordionOpen] = useState<string[]>([]);

  const form = useForm({
    resolver: zodResolver(schema),
    // ... other config
  });

  const onInvalid = (errors: FieldErrors<FormData>) => {
    const sectionsToOpen: string[] = [];
    
    if (personalFields.some(field => errors[field])) {
      sectionsToOpen.push("personal-info");
    }
    if (accountFields.some(field => errors[field])) {
      sectionsToOpen.push("account-details");
    }

    setAccordionOpen(prev => [...prev, ...sectionsToOpen]);
  };

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit, onInvalid)}>
        <Accordion value={accordionOpen} onValueChange={setAccordionOpen}>
          <AccordionItem value="personal-info">
            <AccordionTrigger>Personal Info</AccordionTrigger>
            <AccordionContent forceMount>
              {/* Form fields */}
            </AccordionContent>
          </AccordionItem>
        </Accordion>
      </form>
    </Form>
  );
}
```

adjust `AccordionContent` in `accordion.tsx` like so:
```ts
const AccordionContent = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Content>
>(({ className, children, forceMount, ...props }, ref) => (
  <AccordionPrimitive.Content
    ref={ref}
    className={cn(
      "overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down",
      forceMount && "data-[state=open]:block data-[state=closed]:hidden",
      className,
    )}
    {...props}
  >
    <div className="pb-4 pt-0">{children}</div>
  </AccordionPrimitive.Content>
));
AccordionContent.displayName = AccordionPrimitive.Content.displayName;
```

**Comment by Sam-Kruglov** [comment]
You have to add `forceMount` on the content to make it work. You can't submit because by default when the accordion is closed its content is not even rendered, so it's you don't have the inputs attached to the form.

The old comment below is my old code that I have now removed because I moved to Tanstack Form (highly recommend btw) and so I don't use html form validation anymore, all validation is now handled by the framework instead.

### Old comment
Another approach for nested accordions (to be used together with `forceMount`):
```tsx
<AccordionPrimitive.Trigger ...>
{!selected && <input
              className="w-[8ch] cursor-default border-none p-0 px-0.5 caret-transparent placeholder:text-black focus-visible:outline-none"
              tabIndex={-1}
              form={formId}
              name={`${formId ?? ''}-sneaky-always-invalid-span`}
              type="text"
              required
              placeholder="Category:"
              value={''}
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              onChange={() => {} /* read-only and validated; muting react warning */}
            />}
{selected && ... selected value ... }
</AccordionPrimitive.Trigger>
```

The idea is to make another input tag that cannot be edited and which styled just like a span. Its value is always empty yet it's required, so it's always invalid, its placeholder is what the title should say. When this accordion is closed, the content is not rendered but this input is because it's in the trigger, so when you try to submit the form, this input lights up. I think it's clear enough.

In my usecase I need to select 1 radio button across many which are also grouped into accordions. So, all radio buttons under this accordion have the same input name. So, I don't really need to expand them when invalid hits, I just want to highlight that this category needs to be selected. I just expand the root accordion and show the the heading as invalid. Once the field is selected I rended the heading as regular text.

I still need to use `forceMount` on content together with `props.forceMount && 'data-[state=closed]:hidden'` class on `cn` so that I can collapse the accordion as soon as I made the selection AND keep the actual inputs in the dom so that FormData will work.
